src/compiler/symbolWalker.ts

```diff
@@ -0,0 +1,199 @@
+                visitType(getRestTypeOfSignature(signature));
+                visitType(getReturnTypeOfSignature(signature));
+            }
+
+            function visitInterfaceType(interfaceT: InterfaceType): void {
+                visitObjectType(interfaceT);
+                visitTypeList(interfaceT.typeParameters);
+                visitTypeList(getBaseTypes(interfaceT));
+                visitType(interfaceT.thisType);
+            }
+
+            function visitObjectType(type: ObjectType): void {
+                const stringIndexType = getIndexTypeOfStructuredType(type, IndexKind.String);
+                visitType(stringIndexType);
+                const numberIndexType = getIndexTypeOfStructuredType(type, IndexKind.Number);
+                visitType(numberIndexType);
+
+                // The two checks above *should* have already resolved the type (if needed), so this should be cached
+                const resolved = resolveStructuredTypeMembers(type);
+                for (const signature of resolved.callSignatures) {
+                    visitSignature(signature);
+                }
+                for (const signature of resolved.constructSignatures) {
+                    visitSignature(signature);
+                }
+                for (const p of resolved.properties) {
+                    visitSymbol(p);
+                }
+            }
+
+            function visitSymbol(symbol: Symbol): boolean {
+                if (!symbol) {
+                    return;
+                }
+                const symbolIdString = getSymbolId(symbol).toString();
+                if (visitedSymbols.has(symbolIdString)) {
+                    return;
+                }
+                visitedSymbols.set(symbolIdString, symbol);
+                if (!accept(symbol)) {
+                    return true;
+                }
+                const t = getTypeOfSymbol(symbol);
+                visitType(t); // Should handle members on classes and such
+                if (symbol.flags & SymbolFlags.HasExports) {
+                    symbol.exports.forEach(visitSymbol);
+                }
+                forEach(symbol.declarations, d => {
+                    // Type queries are too far resolved when we just visit the symbol's type
+                    //  (their type resolved directly to the member deeply referenced)
+                    // So to get the intervening symbols, we need to check if there's a type
+                    // query node on any of the symbol's declarations and get symbols there
+                    if ((d as any).type && (d as any).type.kind === SyntaxKind.TypeQuery) {
+                        const query = (d as any).type as TypeQueryNode;
+                        const entity = leftmostSymbol(query.exprName);
+                        visitSymbol(entity);
+                    }
+                });
+            }
+        }
+
+        function leftmostSymbol(expr: QualifiedName | Identifier): Symbol {

```

Comment:
This can probably just be implemented as `getResolvedSymbol(getFirstIdentifier(expr))`, if you're willing to have `getFirstIdentifier` from the checker as a parameter.